affinity/cpp-20/dXXX1r0.md

```diff
@@ -73,25 +75,57 @@ The affinity problem is especially challenging for applications whose behavior c
 
-Consider a code example *(Listing 1)* that uses the C\+\+17 parallel STL algorithm `for_each` to modify the entries of a `valarray` `a`.  The example applies a loop body in a lambda to each entry of the `valarray` `a`, using an execution policy that distributes work in parallel across multiple CPU cores. We might expect this to be fast, but since `valarray` containers are initialized automatically and automatically allocated on the master thread's memory, we find that it is actually quite slow even when we have more than one thread.
+Consider a code example *(Listing 1)* that uses the C\+\+17 parallel STL algorithm `for_each` to modify the entries of a `std::vector` `data`.  The example applies a loop body in a lambda to each entry of the `std::vector` `data`, using an execution policy that distributes work in parallel across multiple CPU cores. We might expect this to be fast, but since `std::vector` containers are initialized automatically and automatically allocated on the master thread's memory, we find that it is actually quite slow even when we have more than one thread.
+
+```cpp
+// NUMA executor representing N NUMA regions.
+numa_executor exec;
+
+// Storage required for vector allocated on construction local to current thread
+// of execution, (N == 0).
+std::vector<float> data(N * SIZE);
+
+// Require the NUMA executor to bind it's migration of memory to the underlying
+// memory resources in a scatter patter.
+auto affinityExec = std::execution::require(exec,
+  bulk_execution_affinity.scatter);
+
+// Migrate the memory allocated for the vector across the NUMA regions in a
+// scatter pattern.
+std::execution::migrate(std::begin(data), std::end(data), affinityExec);
+
+// Placement of data is local to NUMA region 0, so data for execution on other
+// NUMA nodes must is migrated when accessed.
+std::for_each(std::execution::par.on(affinityExec), std::begin(data),
+  std::end(data), [=](float &value) { do_something(value): });
+```
+*Listing 1: Migrating previously allocated memory.*
 
 ```cpp
-// C++ valarray STL containers are initialized automatically.
-// First-touch allocation thus places all of a on the master.
-std::valarray<double> a(N);
-
-// Data placement is wrong, so parallel update is slow.
-std::for_each(std::execution::par, std::begin(a), std::end(a), 
-              [=] (double& a_i) { a_i *= scalar; });
-	      
-// Use future affinity interface to migrate data at next
-// use and move pages closer to next accessing thread.
-...
-// Faster, because data are local now.
-std::for_each(std::execution::par, std::begin(a), std::end(a), 
-              [=] (double& a_i) { a_i *= scalar; });
+// NUMA executor representing N NUMA regions.
+numa_executor exec;
+
+// Reserve space in a vector for a unique_ptr for each index in the bulk
+// execution.
+std::vector<std::unique_ptr<float>> data{};
+data.reserve(N * SIZE);
```

Comment:
I'm a bit confused about this example.  I imagine that this data structure would be a vector of size N, with entries of type `unique_ptr<float[]>`, each of which has size `SIZE`.  The comment "[l]aunch a bulk execution that will allocate each `unique_ptr` in the vector with locality to the nearest NUMA region" seems to suggest this.  Also, the array of arrays data structure would match a PGAS-style explicit affinity programming model.  Is that actually what you meant to do here?